src/pluginextension.c

```diff
@@ -0,0 +1,263 @@
+ * and call @c plugin_extension_register() afterwards.
+ * 
+ * @param extension The @c PluginExtension structure instance provided by the
+ * plugin.
+ * @param ext_name Human-readable name of the extension that can appear in
+ * the user interface. The string should be reasonably unique so extensions can
+ * be distinguished from each other.
+ * @param priority The higher number, the earlier in the extension list the
+ * plugin is and the higher chance it becomes called when multiple extensions
+ * compete for the same feature.
+ * @param data User data passed to the functions from the @c PluginExtension
+ * struct.
+ * @warning Plugins are responsible for calling @c plugin_extension_unregister()
+ * when they no longer provide the extension.
+ *
+ * @see @c plugin_extension_unregister().
+ *
+ * @since 2.1
+ **/
+GEANY_API_SYMBOL
+void plugin_extension_register(PluginExtension *extension, const gchar *ext_name,
+	gint priority, gpointer data)
+{
+	PluginExtensionEntry *entry;
+
+	g_return_if_fail(ext_name != NULL);
+
+	entry = g_malloc(sizeof *entry);
+	entry->extension = extension;
+	entry->data = data;
+	entry->priority = priority;
+
+	all_extensions = g_list_insert_sorted(all_extensions, entry, sort_extension_entries);
+}
+
+
+/**
+ * Plugins are responsible for calling this function when they no longer
+ * provide the extension, at the latest in the plugin @c cleanup() function.
+ * 
+ * @param extension The @c PluginExtension structure that is being unregistered.
+ *
+ * @see @c plugin_extension_register().
+ *
+ * @since 2.1
+ **/
+GEANY_API_SYMBOL
+void plugin_extension_unregister(PluginExtension *extension)
+{
+	for (GList *node = all_extensions; node; node = node->next)
+	{
+		PluginExtensionEntry *entry = node->data;
+
+		if (entry->extension == extension)
+		{
+			g_free(entry);
+			all_extensions = g_list_delete_link(all_extensions, node);
+			break;
+		}
+	}
+}
+
+
+/*
+ * @brief Checks whether a feature is provided
+ * @param f The virtual function name
+ * @param doc The document to check the feature on
+ * @param ext A @c PluginExtension, or @c NULL
+ * @returns @c TRUE if the feature is provided, @c FALSE otherwise.  If @p ext
+ *          is @c NULL, it check whether any extension provides the feature;
+ *          if it is an extension, it check whether it's this extension that
+ *          provides the feature (taking into account possible overrides).
+ */
+#define CALL_PROVIDED(f, doc, ext)												\
+	G_STMT_START {																\
+		for (GList *node = all_extensions; node; node = node->next)				\
+		{																		\
+			PluginExtensionEntry *entry = node->data;							\
+																				\
+			if (entry->extension->f && entry->extension->f(doc, entry->data))	\
+				return (ext) ? entry->extension == (ext) : TRUE;				\
+																				\
+			if ((ext) && entry->extension == (ext))								\
+				return FALSE;													\
+		}																		\
+		return FALSE;															\
+	} G_STMT_END
+
```

Comment:
That's not the whole story - in addition to the static priority, there's also a dynamic aspect of all this.

Consider for instance `autocomplete_provided()` and `autocomplete_perform()` and the LSP plugin and let's say that for the queried document there's a LSP server configured.
1. Geany calls `autocomplete_provided()` of the LSP plugin - since the server isn't started yet, the plugin returns `FALSE` but launches the LSP server in parallel.
2. Until the server is started and the handshake with it completed, other registered extensions may perform the work even if they are behind in the chain.
3. Some time passes and `autocomplete_provided()` is called again by Geany. This time, the server is started. But even though it's started, it doesn't automatically mean it supports autocompletion - so depending on the server capabilities, the plugin returns either `TRUE` and starts performing autocompletion, or `FALSE` and keep others in the chain performing it.
4. If the server is re-configured or e.g. crashes or doesn't run for some other reason, others may perform autocompletion again.
5. And of course, if you disable autocompletion in the LSP config, it will always return `FALSE` without any attempt to start the LSP server.

Anyway, every `_provided()` function should be fast and return quickly without much processing. And I really don't expect thousands of plugins wanting to implement this interface.

Reply:
True, I omitted that Geany will still walk the per extension list until it finds an extension that accepts its request, but at least its only walking a list of the extensions that offer that feature, I would expect most extensions (other than LSP) will only offer one or two features.

But as you say, for now its not critical, but what would be a Geany discussion without a little premature optimisation :grin:?

Reply:
> But as you say, for now its not critical, but what would be a Geany discussion without a little premature optimisation 😁?

Hehe :grin: But really here I think it's best to keep it a bit simpler.  If this *really* becomes an issue, I'm sure you'll come up with an optimized implementation :smile: 
It shouldn't have to change the interface as well, so it can wait.